tests/python/pants_test/task/test_task.py

```diff
@@ -95,15 +105,17 @@ def test_incremental(self):
     vtB = task.execute()
     self.assertContent(vtB, one + two)
 
-    # Incremental atop existing directory for vtA.
+    # vtC.previous_cache_key == vtB.cache_key so the task appends the file to vtB's results.
+    # This behavior used to be skipped because the current_dir existed when checked by cache_manager._use_previous_dir.
```

Comment:
I understood why it didn't copy the previous_results ([the directory was found to exist](https://github.com/pantsbuild/pants/blob/master/src/python/pants/invalidation/cache_manager.py#L203)) but it was not clear to me that was happening on purpose.

And the task itself doesn't noop, which is what I meant above. The invalidation block runs all three times. Zinc just passively inherits the existing results_dir since the namespace is reused. That may be equivalent to a noop for the zinc tool, for all I know, but either way no other tasks even hit that check, since zinc is the only thing that hits that previous_dir code.

Maybe the build_invalidator could go ahead and be fully namespaced in the same manner as the results_dir, and therefore hold an arbitrary number of VT results.  Since those would only be VTs that successfully exited invalidation, the results_dir would only require a wipe if the current_results_dir exists but no hash file was dropped.

I guess that is firmly in the land of a new feature, in any rate, and possibly you know reasons that my speculating is flawed :)

I will be able to provide the contract detailed above. Tasks that are rebuilding cache_keys older than the most recent will consume from the cache, and if not found will be opting out of incremental support since they will no longer passively inherit their results.

Which is maybe preferred when taken as a whole? Cache hits look like they supercede the previous_results, in which case all cached VTs were copying over the previous results in the cache_manager, and then immediately pruning the files within the artifact_cache unpack.